### PR TITLE
【Hackathon 5th No.61】Skillful nowcasting of extreme precipitation with NowcastNet

### PR DESCRIPTION
**Pull request by co63oc**
### PR types

Others 
### PR changes

Others
### Describe

[No.61：Skillful nowcasting of extreme precipitation with NowcastNet](https://github.com/PaddlePaddle/community/blob/master/hackathon/hackathon_5th/%E3%80%90PaddlePaddle%20Hackathon%205th%E3%80%91%E5%BC%80%E6%BA%90%E8%B4%A1%E7%8C%AE%E4%B8%AA%E4%BA%BA%E6%8C%91%E6%88%98%E8%B5%9B%E7%A7%91%E5%AD%A6%E8%AE%A1%E7%AE%97%E4%BB%BB%E5%8A%A1%E5%90%88%E9%9B%86.md#no61skillful-nowcasting-of-extreme-precipitation-with-nowcastnet)

论文链接：
https://www.nature.com/articles/s41586-023-06184-4#Abs1

参考代码链接：
https://codeocean.com/capsule/3935105/tree/v1

RFC https://github.com/PaddlePaddle/community/pull/705

使用预训练模型生成图

torch测试图
image

paddle测试图
image

运行命令
```
python nowcastnet.py CASE_TYPE=normal
python nowcastnet.py CASE_TYPE=large

```

预训练模型和数据集下载
```
链接：https://pan.baidu.com/s/191_Z7IT07RZ86BlK6AHkGA?pwd=x33b 
提取码：x33b 
```

### CODE REVIEW

File: examples/nowcastnet/run.py
Comment (zhiminzhang0830) [inline examples/nowcastnet/run.py:1]: 案例文件需要使用Solver等PaddleScience内置的基础类，这样能与其他案例文件保持统一的风格，因为本案例只需要推理、可视化，因此需要新建Visualizer，可参考开发文档，或 bracket 案例的代码实现

[开发文档](https://paddlescience-docs.readthedocs.io/zh/latest/zh/development/#210)
[solver.visualize()](https://github.com/PaddlePaddle/PaddleScience/blob/484533b4c8f6474d49605a2517945198b178b851/examples/bracket/bracket.py#L527C23-L527C23)

File: ppsci/arch/nowcastnet.py [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+    Args:
+        configs (Dict[str, Any]): The config dictionary.
+
+    Examples:
+        >>> import ppsci
+        >>> configs = {"input_length":9,"total_length":29,"img_height":512,"img_width":512,"img_ch":2,"ngf":32}
+        >>> model = ppsci.arch.NowcastNet(configs)
+    """
+
+    def __init__(self, configs):
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:18]: 模型类初始化参数需要添加 input_keys, output_keys，这样PaddleScience的其他模块调用模型的时能够保证正确运行，可以参考开发文档中的构建 MLP 的例子

https://paddlescience-docs.readthedocs.io/zh/latest/zh/development/#232

File: examples/nowcastnet/convert_checkpoints.py
[the remark this answers is not in the document]
Comment (co63oc) [reply examples/nowcastnet/convert_checkpoints.py:1]: 链接：https://pan.baidu.com/s/191_Z7IT07RZ86BlK6AHkGA?pwd=x33b 
提取码：x33b

File: examples/nowcastnet/mrms_large_case_test.sh
[the remark this answers is not in the document]
Comment (co63oc) [reply examples/nowcastnet/mrms_large_case_test.sh:1]: 已删除

File: examples/nowcastnet/run.py
Comment (co63oc) [reply examples/nowcastnet/run.py:1]: 已修改使用Visualizer

File: ppsci/arch/nowcastnet.py (hunk 1 above)
Comment (co63oc) [reply ppsci/arch/nowcastnet.py]: 已修改参数

File: ppsci/arch/nowcastnet.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+
+    def __init__(
+        self,
+        input_length: int = 9,
+        total_length: int = 29,
+        img_height: int = 512,
+        img_width: int = 512,
+        img_ch: int = 2,
+        ngf: int = 32,
+    ):
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:32]: PaddleScience使用字典管理输入数据、预测数据、真值数据，因此添加 input_keys, output_keys 参数，这两个参数的使用方式可参考[AFNONet](https://paddlescience-docs.readthedocs.io/zh/latest/zh/api/arch/#ppsci.arch.AFNONet)

File: ppsci/arch/nowcastnet.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+
+        self.pred_length = self.total_length - self.input_length
+        self.evo_net = Evolution_Network(self.input_length, self.pred_length, base_c=32)
+        self.gen_enc = Generative_Encoder(self.total_length, base_c=self.ngf)
+        self.gen_dec = Generative_Decoder(configs)
+        self.proj = Noise_Projector(self.ngf)
+        sample_tensor = paddle.zeros(shape=[1, 1, self.img_height, self.img_width])
+        self.grid = make_grid(sample_tensor)
+
+    def forward(self, data):
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:67]: PaddleScience中在forward函数内会对输入、输出数据做变换以及将不同的输入数据进行concat, 而将实际的网络前向逻辑放到forward_tensor函数内，因此需要将函数名修改为 forward_tensor，并添加forward函数，参考[AFNONet](https://paddlescience-docs.readthedocs.io/zh/latest/zh/api/arch/#ppsci.arch.AFNONet)

File: ppsci/arch/nowcastnet.py [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+        all_frames = x[:, :, :, :, :1]
+        frames = all_frames.transpose(perm=[0, 1, 4, 2, 3])
+        batch = frames.shape[0]
+        height = frames.shape[3]
+        width = frames.shape[4]
+        # Input Frames
+        input_frames = frames[:, : self.input_length]
+        input_frames = input_frames.reshape((batch, self.input_length, height, width))
+        # Evolution Network
+        intensity, motion = self.evo_net(input_frames)
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:100]: 发现推理结果的图片与参考代码的有些差异，应该是某处没有对齐导致的，我这边帮忙查了下发现这个网络给定相同输入时输出存在2e-4左右的差异，需要在详细查一下，修改后建议再对整个网络整体检查下对齐情况

Comment (co63oc) [reply ppsci/arch/nowcastnet.py:100]: 初始化随机数不同结果有差异

Comment (luotao1) [reply ppsci/arch/nowcastnet.py:100]: 当前情况：paddle的grid_sample的CPU结果跟torch的GPU结果能对齐，GPU结果跟torch的CPU结果能对齐，但是把这个运算改到cpu上，后面仍然有没对齐的地方。by @HydrogenSulfate @zhiminzhang0830 
- https://github.com/PaddlePaddle/Paddle/issues/58907

Comment (zhiminzhang0830) [reply ppsci/arch/nowcastnet.py:100]: > 后面仍然有没对齐

@luotao1 @HydrogenSulfate 这边check过了，后面没对齐仍然是grid_sample的问题，这个PR除了grid_sample外基本OK了，还有几处格式问题需要辛苦 @co63oc 再优化一下就可以合入了。

File: ppsci/arch/nowcastnet.py [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+        self,
+        input_keys: Tuple[str, ...],
+        output_keys: Tuple[str, ...],
+        input_length: int = 9,
+        total_length: int = 29,
+        image_height: int = 512,
+        image_width: int = 512,
+        image_ch: int = 2,
+        ngf: int = 32,
+        num_timestamps: int = 1,
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:37]: NowCastNet 不需要 num_timestamps 参数，可以直接去掉

File: ppsci/arch/nowcastnet.py [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+    def forward(self, x):
+        x = self.inc(x)
+        x = self.down1(x)
+        x = self.down2(x)
+        x = self.down3(x)
+        return x
+
+
+class Generative_Decoder(paddle.nn.Layer):
+    def __init__(self, opt):
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:322]: Generative_Decoder 可以直接改为使用多个入参么，类似这样
`def __init__(self, ngf, evo_ic, gen_oc, ic_feature):`

File: ppsci/arch/nowcastnet.py [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+        actv = self.mlp_shared(evo)
+        gamma = self.mlp_gamma(self.pad(actv))
+        beta = self.mlp_beta(self.pad(actv))
+        out = normalized * (1 + gamma) + beta
+        return out
+
+
+class Noise_Projector(paddle.nn.Layer):
+    def __init__(self, input_length):
+        super(Noise_Projector, self).__init__()
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:449]: 修改为`super().__init__()`，其他类似的地方也需要修改一下

File: ppsci/arch/nowcastnet.py [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+    return grid
+
+
+def warp(input, flow, grid, mode="bilinear", padding_mode="zeros"):
+    B, C, H, W = input.shape
+    vgrid = grid + flow
+    vgrid[:, 0, :, :] = 2.0 * vgrid[:, 0, :, :].clone() / max(W - 1, 1) - 1.0
+    vgrid[:, 1, :, :] = 2.0 * vgrid[:, 1, :, :].clone() / max(H - 1, 1) - 1.0
+    vgrid = vgrid.transpose(perm=[0, 2, 3, 1])
+    output = paddle.nn.functional.grid_sample(
```
Comment (zhiminzhang0830) [inline ppsci/arch/nowcastnet.py:527]: 由于grid_sample无法对齐的问题，可以暂时在此处使用cpu进行计算
```
    output = paddle.nn.functional.grid_sample(
        x=input.cpu(), grid=vgrid.cpu(), padding_mode=padding_mode, mode=mode, align_corners=True
    )
    output = output.cuda()
```

File: ppsci/visualize/radar.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        super().__init__(input_dict, output_expr, batch_size, num_timestamps, prefix)
+        self.case_type = case_type
+        self.total_length = total_length
+        self.input_dict = input_dict
+
+    def save(self, path, data_dict):
+        if not os.path.exists(path):
+            os.makedirs(path)
+        test_ims = self.input_dict[list(self.input_dict.keys())[0]]
+        img_gen = data_dict[list(data_dict.keys())[0]]
```
Comment (zhiminzhang0830) [inline ppsci/visualize/radar.py:74]: data_dict中有input、output两个key，此处需要取出output key下的数据，需要修改为以下代码，否则会将label数据当作网络预测结果画到图上
`img_gen =  data_dict[self.output_keys[0]]`

File: ppsci/visualize/radar.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            test_ims_plot = test_ims[0][
+                :-2, 256 - 192 : 256 + 192, 256 - 192 : 256 + 192
+            ]
+            img_gen_plot = img_gen[0][:-2, 256 - 192 : 256 + 192, 256 - 192 : 256 + 192]
+        else:
+            test_ims_plot = test_ims[0][:-2]
+            img_gen_plot = img_gen[0][:-2]
+        save_plots(
+            test_ims_plot,
+            labels=["gt{}".format(i + 1) for i in range(self.total_length)],
```
Comment (zhiminzhang0830) [inline ppsci/visualize/radar.py:86]: PaddleScience中使用f-string，可以修改为以下格式：
`labels=[f"gt{i + 1)}" for i in range(self.total_length)],`

File: ppsci/visualize/radar.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            res_path=path,
+            vmin=vis_info["vmin"],
+            vmax=vis_info["vmax"],
+        )
+        save_plots(
+            img_gen_plot,
+            labels=["pd{}".format(i + 1) for i in range(9, self.total_length)],
```
Comment (zhiminzhang0830) [inline ppsci/visualize/radar.py:93]: 同上，修改为f-string

**Comment by zhiminzhang0830** [comment]
NowCastNet需要用到 OpenCV, 需要在requirements.txt里添加该库 opencv-python

File: ppsci/arch/nowcastnet.py (hunk 6 above)
Comment (co63oc) [reply ppsci/arch/nowcastnet.py:316]: 其他子层需要使用opt，如果拆开其他子层也要拆开，代码参数变量较多

**Comment by co63oc** [comment]
> NowCastNet需要用到 OpenCV, 需要在requirements.txt里添加该库 opencv-python

已修改

File: ppsci/visualize/radar.py (hunk 9 above)
Comment (co63oc) [reply ppsci/visualize/radar.py]: 修改为 img_gen = data_dict[list(data_dict.keys())[1]]，1是output

File: docs/zh/api/data/dataset.md [hunk 12]
```diff
@@ -20,5 +20,6 @@
         - VtuDataset
         - MeshAirfoilDataset
         - MeshCylinderDataset
+        - RadarDataset
```
Comment (zhiminzhang0830) [inline docs/zh/api/data/dataset.md:23]: 同样地，在arch.md和visualize.md添加 NowcastNet 和 VisualizerRadar

File: docs/zh/examples/nowcastnet.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+--8<--
+```
+
+然后构建 VisualizerRadar 生成图片结果：
+
+``` py linenums="69" title="examples/nowcastnet/nowcastnet.py"
+--8<--
+examples/nowcastnet/nowcastnet.py:69:82
+--8<--
+
```
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet.md:70]: 此处少了 代码的结束符号 ```

File: docs/zh/examples/nowcastnet.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+=== "模型评估命令"
+
+    ``` sh
+    # linux
+    wget https://paddle-org.bj.bcebos.com/paddlescience/datasets/nowcastnet/nowcastnet.zip
+    # windows
+    # curl https://paddle-org.bj.bcebos.com/paddlescience/datasets/nowcastnet/nowcastnet.zip --output nowcastnet.zip
+    unzip nowcastnet.zip -d datasets/
+    python nowcastnet.py mode=eval EVAL.pretrained_model_path=https://paddle-org.bj.bcebos.com/paddlescience/models/nowcastnet/nowcastnet_pretrained.pdparams
+    ```
```
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet.md:16]: 数据集和模型已经上传，需要修改下下载路径：
``` sh
    # linux
    wget https://paddle-org.bj.bcebos.com/paddlescience/datasets/nowcastnet/mrms.tar
    # windows
    # curl https://paddle-org.bj.bcebos.com/paddlescience/datasets/nowcastnet/mrms.tar --output mrms.tar
    tar -xvf mrms.tar -C datasets/
    python nowcastnet.py mode=eval EVAL.pretrained_model_path=https://paddle-org.bj.bcebos.com/paddlescience/models/nowcastnet/nowcastnet_pretrained.pdparams
```

File: docs/zh/examples/nowcastnet.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+## 2. 模型原理
+
+本章节仅对 NowcastNet 的模型原理进行简单地介绍，详细的理论推导请阅读 [Skilful nowcasting of extreme precipitation with NowcastNet](https://www.nature.com/articles/s41586-023-06184-4#Abs1)。
+
+模型的总体结构如图所示：
+
+<figure markdown>
+  ![nowcastnet-arch](nowcastnet/nowcastnet.png){ loading=lazy style="margin:0 auto"}
+  <figcaption>NowcastNet 网络模型</figcaption>
+</figure>
```
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet.md:32]: 图片已上传，可修改改为：

```
<figure markdown>
  ![nowcastnet-arch](https://paddle-org.bj.bcebos.com/paddlescience/docs/nowcastnet/nowcastnet.png){ loading=lazy style="margin:0 auto"}
  <figcaption>NowcastNet 网络模型</figcaption>
</figure>
```

File: docs/zh/examples/nowcastnet.md [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+
+下图展示了模型的预测结果和真值结果。
+
+<figure markdown>
+  ![result](nowcastnet/pd.gif){ loading=lazy style="margin:0 auto;"}
+  <figcaption>模型预测结果</figcaption>
+</figure>
+
+<figure markdown>
+  ![result](nowcastnet/gt.gif){ loading=lazy style="margin:0 auto;"}
```
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet.md:89]: 图片地址替换为：https://paddle-org.bj.bcebos.com/paddlescience/docs/nowcastnet/gt.gif

File: docs/zh/examples/nowcastnet.md [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+examples/nowcastnet/nowcastnet.py
+--8<--
+```
+
+## 6. 结果展示
+
+下图展示了模型的预测结果和真值结果。
+
+<figure markdown>
+  ![result](nowcastnet/pd.gif){ loading=lazy style="margin:0 auto;"}
```
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet.md:84]: 图片已上传，地址替换为：https://paddle-org.bj.bcebos.com/paddlescience/docs/nowcastnet/pd.gif

File: docs/zh/examples/nowcastnet/gt.gif
Comment (zhiminzhang0830) [inline docs/zh/examples/nowcastnet/gt.gif:1]: 此文件夹下的内容已上传，可以删除此文件夹

File: ppsci/visualize/radar.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+        ...     "y": np.random.randn(128, 1),
+        ...     "u": np.random.randn(128, 1),
+        ...     "v": np.random.randn(128, 1),
+        ... }
+        >>> visualizer_u_v =  ppsci.visualize.VisualizerRadar(
+        ...     vis_points,
+        ...     {"u": lambda d: d["u"], "v": lambda d: d["v"]},
+        ...     num_timestamps=1,
+        ...     prefix="result_u_v",
+        ... )
```
Comment (zhiminzhang0830) [inline ppsci/visualize/radar.py:52]: 该例子需要尽量接近实际情况。对于input_dict 中的数据可以随机初始化一个与实际情况shape一样的tensor，对于output_expr可以使用案例代码中的写法：
```
{"output": lambda out: out["output"]}
```

File: docs/zh/api/data/dataset.md (hunk 12 above)
Comment (co63oc) [reply docs/zh/api/data/dataset.md:23]: 已修改

**Review by zhiminzhang0830 — APPROVED** [review]
LGTM
